### ISSUE TITLE
[C++] Add tests against analytically computable cases to GP mean, variance computation

### ISSUE THREAD

**Issue by suntzu86** [Yelp/MOE#157]
The testing of KrigingBeliever (gpp_heuristic_expected_improvement_test.cpp) uses the fact that we can analytically know the result of GP (with 0 noise) mean, variance (and their gradients) in certain special cases. In particular,
1. If evaluated at training points (i.e., the points in X_s are a subset of the points in X), the output will be the function values observed at those points (i.e., "f") with variance = 0.
2. If evaluated "very far" away ('very far' can be estimated; see gpp_heuristic_expected_improvement_test.cpp), the GP mean is 0 and the variance is \alpha, the signal_variance hyperparameter.

These should be tested against single and multiple points_to_sample (X_s).

These cases provide some extra sanity checks on the current ping testing framework and are worth implementing. It should be very simple to do so b/c KrigingBelieverPolicyTest() already uses many of these facts.
